sdk/identity/azure-identity/src/managed_identity_source.cpp

```diff
@@ -482,10 +482,10 @@ std::unique_ptr<ManagedIdentitySource> ImdsManagedIdentitySource::Create(
 ImdsManagedIdentitySource::ImdsManagedIdentitySource(
     std::string const& clientId,
     std::string const& resourceId,
-    Azure::Core::Url const& imdsUrl,
+    Azure::Core::Url imdsUrl,
```

Comment:
The [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in) say that all non-POD parameters should be passed by const reference because passing by value forces the caller to construct the copy and destroy the copy.

Passing by const& means that the copy is made in the constructor which can dramatically reduce the generated code size of the module.

Reply:
Line 45 vs line 46 is what I am talking about: https://godbolt.org/z/TKEKnfrr6

Reply:
I'm not sold, and the guidelines are unambiguous: you should not pass non-POD types by value to constructors or other methods.

Consider [this variant of your code](https://godbolt.org/z/1M593MYKz): Note that there are TWO constructor calls for Attribute, and two destructor calls. The first constructor call creates the original object, the second is an invocation of the copy constructor.

The second constructor is completely unnecessary if the value is passed by const reference.

The only time that passing the value by non-const reference is a win is the example you linked, where you construct the object and immediately pass it into the constructor. But unless you can *enforce* that pattern (and I don't know how you can do that), then most uses of this are inefficient.

Reply:
This is exactly my point! Since class `C` is going to be keeping a copy of an object anyways, the fact that it happens during parameter construction makes no difference!

Your code, output with byref (line 45 commented, 46 uncommented):
```
strlen
allocate buffer
copy to buffer
output to make it clear
strlen
allocate buffer
copy to buffer
deallocate buffer
deallocate buffer
```

Your code, output with byval+move (line 45 uncommented, 46 commented):
```
strlen
allocate buffer
copy to buffer
output to make it clear
strlen
allocate buffer
copy to buffer
transfer buffer ownership <-- this is extra
deallocate buffer
deallocate buffer
```

i.e. exactly the same in terms of heavy operations, except for the buffer ownership transfer, but it is an inexpensive operation!

But now, in case the parameters were created in-place, byval+move is more efficient.
Original example, output with byref (line 45 commented, 46 uncommented):
```
strlen
allocate buffer
copy to buffer
strlen
allocate buffer
copy to buffer
deallocate buffer
deallocate buffer
```

Original example, output with byval+move (line 45 uncommented, 46 commented):
```
strlen
allocate buffer
copy to buffer
transfer buffer ownership
deallocate buffer
```

Reply:
How would our guideline here impact code as it evolves and the implementation changes? Code and implementation is rarely static and we don't want guidelines to be dependent on how something looks **today**.
If someone makes a change to the underlying parameter (if it gets copied/stored where it wasn't before), would they be expected to update the various signatures too? That seems difficult to adhere to.

It is important to consider a few things when it comes to design guidelines:
1) How easy are they to understand/follow, and verify they are being followed by contributors and code reviewers. 
2) They should be consistently applicable, so they avoid re-hashing the same conversation. That's important since such guidelines improve code velocity and reduce review friction.
3) They should be focused on the public surface area and be resilient to code/implementation changes, as it evolves. If guidelines are implementation detail dependent, then they can't be as stable.

Reply:
It is not a breaking change to change `C(Attribute attr)` to `C(const Attribute& attr)` or vice versa. As you can see, we can comment/uncomment lines 45/46 without modifying the code in `main()`. So, lines 45/46 represent our SDK, and the `main()` represents the customer's code, and it keeps compiling and nothing breaks no matter which direction you make the change.

Reply:
When invoking as `C c{Attribute("Hello")}`, my code performs better: memory is allocated once. With your code, memory is allocated twice.

When invoking as `Attribute a{"Hello"}; C c{a};`, my code and your code perform similarly: memory is allocated twice. The only extra thing that my code does is that it updates two extra pointers, but that is negligible relative to strlen+malloc+strcpy.

Am I missing something?

Reply:
And you want it to always allocate buffers twice, to avoid paying the price of updating two pointers in a move constructor, correct?

Reply:
I want to follow the best practices as expressed in the C++ core guidelines. And those guidelines are unambiguous: Pass non-trivial values by const reference.

This is a case where we should follow the guidelines unless there is an unambiguous reason for not following them.

And looking at the generated code, I see a copy constructor being invoked in the by value version that isn't invoked in the by reference version. 

The by reference version does not invoke the copy constructor to create a temporary - it just passes the value in by reference.

The *only* case where your code performs better is when you are explicitly creating the temporary in the constructor invocation. If the object being passed into the constructor is an existing object, the performance is worse because passing by value forces the creation of a temporary object.

Reply:
> the performance is worse because passing by value forces the creation of a temporary object.
But all that temporary object holds ends up being moved into `C::m_attribute`, so all that work being done for the temporary object was not for nothing. When we take byref and do the copying ourselves, we are doing exactly the same work that the temporary object does when it gets created.

I knowingly started this discussion with an expectation that we can discuss beyond the current guidelines, as maybe it can get amended as an outcome of this discussion.

We could also ask Billy or on specific aliases, but honestly I am not that much passionate about this - all I did was that I saw the pattern that could be better in my opinion, and suggested to change it for something slightly better in my opinion. I did not block the PR. FWIW, I wrote most of my code for this SDK with byval+move. Prior to C++11, I was using crefs+copy ctors, as there was no better alternative. Most likely, the most performant thing would be to have two overloads - one for `T const&`, and one for `T&&`, but that is also twice the code to maintain, so I think of byval+move as some tradeoff in between.

Anyways, I think I have said everything I could say, and if we continue, we'll just go in circles, so please do what you want with this PR; even if I have to use crefs+copy ctors from now on, I am ok with that.

Reply:
I would also guess that `Attribute a{"Hello"}; C c{a};` and similar situations, where `a` is only used as a parameter to `c`, gets optimized into `C c{Attribute("Hello")}` more often than we can get an impression from just reading the code.

Reply:
Apparently my response got lost.

I want our code to conform to the C++ Core Guidelines whenever possible, and when it doesn't, I want there to be an explicit section in the guidelines that explains why we're deviating from the core guidelines (simple example of a deviation from the guidelines: Our methods and classes are CamelCase, not snake_case).

The C++ Core Guidelines explicitly state that all non-trivial parameters should be passed by const reference rather than by value, and they lay out a number of reasons for that choice.

In this case as I commented in my email, the "copy" variant of the constructor has to call the copy constructor for the object to create a temporary copy of the object, which is then subsequently moved. The "const &" variant of the constructor doesn't have to create a temporary copy of the object, instead the constructor calls the copy constructor.

The key difference is in the call site, NOT the allocation pattern - if we assume that every parameter to the constructor is constructed via a temporary (`C c{Attribute("hello"))`), then the call site performance is identical. If we assume that some parameters are not constructed via temporary (`Attribute a{"hello"}; Cc{a}`), then the compiler will generate a copy constructor and a destructor for each call site, which results in unnecessary code bloat.

Typically, there are far more calls to a method than there are methods. And that's why passing non-trivial objects by const reference matters - it generates more code.

So as I mentioned at the outset: The current C++ guidelines incorporate the C++ Core Guidelines by reference. We need to document the methodology behind all variations from those guidelines.

Reply:
Btw, I annotated the example to show differences between the patterns.

First consider the case where the parameter is an existing value:

`CopyParam c{attribute}`:
```
strlen
allocate buffer
copy to buffer<= Allocate a temporary copy of attribute.
transfer buffer ownership <= Move the temporary to C.
```

`ReferenceParam c{attribute}`
```
strlen
allocate buffer
copy to buffer <= Copy the attribute to C.
```

These two are essentially identical in performance, assuming that moving the temporary is essentially free.


Next consider the temporary parameter case:

`CopyParam(Attribute)`
```
strlen
allocate buffer
copy to buffer <= This is creating the temporary.
transfer buffer ownership <= Move the temporary to C.
```

`ReferenceParam(Attribute)`
```
strlen
allocate buffer
copy to buffer <= Create the temporary.
strlen
allocate buffer
copy to buffer <= Copy the temporary to C.
```

In this case, the by reference parameter IS worse, because of the creation of the temporary.

So the only case where the performance is better with the "value parameter" case is when you are constructing a temporary object on the stack. In all other cases the performance is identical.

But the cost of this performance optimization is hidden code generated at every call site, bloating the size of the generated binary.

I'm not sure we can guarantee that our customers will always construct the parameter as a temporary, and the cost of passing by value is additional code at every call site.

I'm not sure that this is worth doing.

Reply:
I think we often don't create parameters inside constructor parenthesis for the sole reason that the code is more readable if parameter values are initialized on a separate line before invoking the constructor. But essentially after being passed to a constructor, the value is never referenced. So when the compiler sees
```cpp
Widget GetWidget(int a, int b, int c) {
    WidgetParam left{a, b+c, a+b};
    WidgetParam center{b, a+c, a+a}; // a comment explaining what is going on
    WidgetParam right{c, b, a}; // Plus, WidgetParams are given names making it easier to read

    Widget result{left, center, right};
    // left, center, right, are not used amymore
    return result;
}
```
it may optimize it to
```cpp
Widget GetWidget(int a, int b, int c) {
    return Widget{WidgetParam{a, b+c, a+b}, WidgetParam{b, a+c, a+a}, WidgetParam{c, b, a}};
}
```

In which case there will be two allocations in const ref case, but one allocation and one ownership transfer in byval+move case.